{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.  Convert BAM into SAM to view\n",
    "\n",
    "    ```\n",
    "    samtools view -h -o aln.sam aln.bam\n",
    "    ```\n",
    "    \n",
    "2.  View file.  We see that the file is unsorted.  \n",
    "\n",
    "3.  Sort BAM file\n",
    "\n",
    "    ```\n",
    "    samtools sort aln.bam > alnsorted.bam\n",
    "    ```\n",
    "\n",
    "4.  Let's reindex\n",
    "\n",
    "    ```\n",
    "    samtools index alnsorted.bam > alnsorted.bam.bai\n",
    "    ```\n",
    "    \n",
    "5.  Let's call variants using bcftools\n",
    "\n",
    "    ```\n",
    "    bcftools mpileup -Ou -f ./../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa ./../../notebooks/data/samtools_output/aln_sorted.bam | bcftools call -vmO z -o aln_calls.vcf.gz\n",
    "    ```\n",
    "    \n",
    "    **This didn't work.**\n",
    "    \n",
    "    \n",
    "6.  Let's also call variants using freebayes\n",
    "\n",
    "    ```\n",
    "    freebayes --fasta-reference ./../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa ./../../notebooks/data/samtools_output/aln_sorted.bam > ./../../notebooks/data/freebayes_output/var.vcf\n",
    "    ```\n",
    "    \n",
    "7.  Zip and index VCF\n",
    "\n",
    "    ```\n",
    "    bgzip var.vcf # makes var.vcf.gz\n",
    "    tabix -p vcf var.vcf.gz\n",
    "    ```\n",
    "    \n",
    "8.  prepare graphs and statistics to assist you in filtering your variants\n",
    "\n",
    "    ```\n",
    "    bcftools stats -F ./../../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa -s - var.vcf.gz > var.vcf.gz.stats\n",
    "    mkdir plots\n",
    "    plot-vcfstats -p plots/ var.vcf.gz.stats\n",
    "    ```\n",
    "    \n",
    "7.  Ploted VCF file using CircosVCF.  Seems like a pretty messy file.  Needs to be cleaned up.  But how.?  We need to take a look at these alignments.  \n",
    "\n",
    "8.  So it looks like the BAM file is garbage.\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__To Index Reference__:\n",
    "\n",
    "```\n",
    "bwa index reference.fa\n",
    "```\n",
    "\n",
    "__To Sort BAM__:\n",
    "\n",
    "```bash\n",
    "samtools sort ./../../notebooks/data/original_files/aln.bam > ./../../notebooks/data/samtools_output/aln_sorted.bam\n",
    "```\n",
    "\n",
    "__To Index BAM__:\n",
    "\n",
    "```bash\n",
    "samtools index ./../../notebooks/data/samtools_output/aln_sorted.bam ./../../notebooks/data/samtools_output/aln_sorted.bam.bai\n",
    "```\n",
    "\n",
    "__To Align__:\n",
    "\n",
    "```\n",
    "bwa mem -M -t 2 ./../../Reference_Sequence/Homo_sapiens.GRCh37/chromosomes/Homo_sapiens.GRCh37.dna_sm.chromosome.3.fa ./../../notebooks/data/original_files/aln1.fastq ./../../notebooks/data/original_files/aln1.fastq > ./../../notebooks/data/bwa_output/aln.sam\n",
    "```\n",
    "\n",
    "__Call Variants__:\n",
    "\n",
    "```\n",
    "freebayes --fasta-reference ./../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa ./../../notebooks/data/samtools_output/aln_sorted.bam > ./../../notebooks/data/freebayes_output/var.vcf\n",
    "```\n",
    "\n",
    "\n",
    "```\n",
    "freebayes --fasta-reference ./../../Reference_Sequence/Homo_sapiens.GRCh37/chromosomes/Homo_sapiens.GRCh37.dna_sm.chromosome.3.fa ./../../notebooks/data/samtools_output/aln_sorted.bam > ./../../notebooks/data/freebayes_output/var_chrm3.vcf\n",
    "```\n",
    "\n",
    "```\n",
    "bcftools mpileup -Ou -d 1150004 -f ./../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa  ./../../notebooks/data/samtools_output/aln_sorted.bam | bcftools call -mv -Ob -o ./../../notebooks/data/bcftools_output/var150004.bcf\n",
    "```\n",
    "\n",
    "```\n",
    "bcftools mpileup -d 1150004 -f ./../../Reference_Sequence/Homo_sapiens.GRCh37/Homo_sapiens.GRCh37.dna_sm.toplevel.fa  ./../../notebooks/data/samtools_output/aln_sorted.bam | bcftools call -mv -Oz -o ./../../notebooks/data/bcftools_output/var1150004.vcf\n",
    "```\n",
    "\n",
    "\n",
    "__Call Filtering__:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So it seems we're going to have to start over from the alignment.\n",
    "\n",
    "So let's do a quick QC on the FASTQ with trimmomatic and FASTQC.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": ".venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
